File: 주영/Tree/113. Path Sum II.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function (root, targetSum) {
  if (!root) return [];
  let result = [];
  const recurPathSum = (node, target, path) => {
    if (!node.left && !node.right && target === node.val) {
      // console.log('path: ', path);
      result.push(path);
    }

    if (node.left)
      recurPathSum(node.left, target - node.val, [...path, node.left.val]);
    if (node.right)
      recurPathSum(node.right, target - node.val, [...path, node.right.val]);
  };
  recurPathSum(root, targetSum, [root.val]);
  return result;
};
let root = [5, 4, 8, 11, null, 13, 4, 7, 2, null, null, 5, 1];
let targetSum = 22;

pathSum(root, targetSum);

File: P/113_Path_Sum_II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>> ans;
    int target;
    
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        if(root == null)
            return new ArrayList();

        this.ans = new ArrayList();
        this.target = targetSum;
        
        findPathSum(root, 0, new ArrayList());
        return ans;
    }
    
    private void findPathSum(TreeNode node, int sum, ArrayList<Integer> arr) {
        sum += node.val;
        arr.add(node.val);
        
        if(node.left == null && node.right == null) {
            if(sum == this.target)
                this.ans.add((List<Integer>)arr.clone());
        } else {
            if(node.left != null)
                findPathSum(node.left, sum, arr);
            if(node.right != null)
                findPathSum(node.right, sum, arr);
        }
        arr.remove(arr.size() - 1);
        return;
    }
}